cmd/provisioner-localpv/app/config.go

```diff
@@ -0,0 +1,217 @@
+	//storagev1 "k8s.io/api/storage/v1"
+)
+
+const (
+	//KeyPVStorageType defines if the PV should be backed
+	// a hostpath ( sub directory or a storage device)
+	KeyPVStorageType = "StorageType"
+	//KeyPVBasePath defines base directory for hostpath volumes
+	// can be configured via the StorageClass annotations.
+	KeyPVBasePath = "BasePath"
+	//KeyPVRelativePath defines the alternate folder name under the BasePath
+	// By default, the pv name will be used as the folder name.
+	// KeyPVBasePath can be useful for providing the same underlying folder
+	// name for all replicas in a Statefulset.
+	// Will be a property of the PVC annotations.
+	KeyPVRelativePath = "RelativePath"
+	//KeyPVAbsolutePath specifies a complete hostpath instead of
+	// auto-generating using BasePath and RelativePath. This option
+	// is specified with PVC and is useful for granting shared access
+	// to underlying hostpaths across multiple pods.
+	KeyPVAbsolutePath = "AbsolutePath"
+)
+
+const (
+	// Some of the PVCs launched with older helm charts, still
+	// refer to the StorageClass via beta annotations.
+	betaStorageClassAnnotation = "volume.beta.kubernetes.io/storage-class"
+)
+
+//getVolumeConfig creates a new VolumeConfig struct by
+// parsing and merging the configuration provided in the PVC
+// annotation - cas.openebs.io/config with the
+// default configuration of the provisioner.
+func (p *Provisioner) GetVolumeConfig(pvName string, pvc *v1.PersistentVolumeClaim) (*VolumeConfig, error) {
+
+	pvConfig := p.defaultConfig
+
+	//Fetch the SC
+	scName := GetStorageClassName(pvc)
+	sc, err := p.kubeClient.StorageV1().StorageClasses().Get(*scName, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get storageclass: missing sc name {%v}", scName)
+	}
+
+	// extract and merge the cas config from storageclass
+	scCASConfigStr := sc.ObjectMeta.Annotations[string(mconfig.CASConfigKey)]
+	glog.Infof("SC %v has config:%v", *scName, scCASConfigStr)
+	if len(strings.TrimSpace(scCASConfigStr)) != 0 {
+		scCASConfig, err := cast.UnMarshallToConfig(scCASConfigStr)
+		if err == nil {
+			pvConfig = cast.MergeConfig(scCASConfig, pvConfig)
+		}
+	}
+
+	//TODO : extract and merge the cas volume config from pvc
+	//This block can be added once validation checks are added
+	// as to the type of config that can be passed via PVC
+	//pvcCASConfigStr := pvc.ObjectMeta.Annotations[string(mconfig.CASConfigKey)]
+	//if len(strings.TrimSpace(pvcCASConfigStr)) != 0 {
+	//	pvcCASConfig, err := cast.UnMarshallToConfig(pvcCASConfigStr)
+	//	if err == nil {
+	//		pvConfig = cast.MergeConfig(pvcCASConfig, pvConfig)
+	//	}
+	//}
+
+	pvConfigMap, err := cast.ConfigToMap(pvConfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to read volume config: pvc {%v}", pvc.ObjectMeta.Name)
+	}
+
+	c := &VolumeConfig{
+		pvName:  pvName,
+		pvcName: pvc.ObjectMeta.Name,
+		scName:  *scName,
+		options: pvConfigMap,
+	}
+	return c, nil
+}
```

Comment:
I can think of VolumeConfig as an object. ( We can try to use following inside `GetVolumeConfig`)
```
// pvName:= via passed argument
// pvcName: = (via pvc package ) pvc.GetPVCName(pvcObj) ( `pvc` is the package)  
// similary other variables can be separately fetched and pushed to builder predicate.
// FYI -- Functions starting with with are predicates. 
vc,err:=NewVolumeConfigBuilder.WithPvName(pvName).WithPvcName(pvcName).WithOptions(options).Build()
```
We can again choose to improve it later.

Reply:
See if following gives some idea. Implementation may differ a/q to your choice

```
package main

type VolumeConfigBuilder struct {
	VolumeConfig   *VolumeConfig
	errs           []error
	validationErrs []error
	fastFail       bool
}

func NewVolumeConfigBuilder() {
	return &VolumeConfigBuilder{
		VolumeConfig: &VolumeConfig{},
	}
}
func (vcb *VolumeConfigBuilder) Build() *VolumeConfig {
        // For exmaple -- if builder predicates are prone to error
        // If len(err)>0 , return nil,err ...
	return vcb.VolumeConfig

}
// Functions starting with With are builder predicates.
// If these can return error , put error return type in predicate and also in above build function
func (vcb *VolumeConfigBuilder) WithPVName(pvName string) *VolumeConfigBuilder {
	// For exmaple -- if builder predicates are prone to error
	
	// if fastFail 
	// return vcb(object),nil(error)
	
	
	// err received
	// vcb.errs=append(vcb.errs,err)
	
	// Similarly in other predicates
	
	vcb.VolumeConfig.pvName = pvName
	return vcb
}

func (vcb *VolumeConfigBuilder) WithPVCName(pvcName string) *VolumeConfigBuilder {
	vcb.VolumeConfig.pvcName = pvcName
	return vcb
}

func (vcb *VolumeConfigBuilder) WithSCName(scName string) *VolumeConfigBuilder {
	vcb.VolumeConfig.pvName = scName
	return vcb
}

func (vcb *VolumeConfigBuilder) WithOptions(options optionsType) *VolumeConfigBuilder {
	vcb.VolumeConfig.options = options
	return vcb
}

func (p *Provisioner) GetPvConfigMap(scName string) (*VolumeConfig, error) {
	pvConfig := p.defaultConfig

	//Fetch the SC

	scObject, err := sc.NewKubeClient.Get(scName, metav1.GetOptions{})
	if err != nil {
		return nil, errors.Wrapf(err, "failed to get storageclass: missing sc name {%v}", scName)
	}

	// extract and merge the cas config from storageclass
	// ALSO THIS LOGIC CAN BE POSSIBLY MOVED TO ANOTHER FUNCTION
	
	scCASConfigStr := sc.ObjectMeta.Annotations[string(mconfig.CASConfigKey)]
	glog.Infof("SC %v has config:%v", *scName, scCASConfigStr)
	if len(strings.TrimSpace(scCASConfigStr)) != 0 {
		scCASConfig, err := cast.UnMarshallToConfig(scCASConfigStr)
		if err == nil {
			pvConfig = cast.MergeConfig(scCASConfig, pvConfig)
		}
	}

	pvConfigMap, err := cast.ConfigToMap(pvConfig)
	if err != nil {
		return nil, errors.Wrapf(err, "unable to read volume config: pvc {%v}", pvc.ObjectMeta.Name)
	}
}

func (p *Provisioner) GetVolumeConfig(pvName string, pvcObject *v1.PersistentVolumeClaim) (*VolumeConfig, error) {
	pvcName := pvc.GetName(pvcObject)
	scName := pvc.GetSCName(pvcObject)
	pvConfigMap := p.GetPVConfigMap(scName, pvcObject)
	c := NewVolumeConfigBuilder.WithPVName().WithPVCName().WithSCName().WithOptions().Build
	return c, nil
}
```

Reply:
https://github.com/openebs/maya/issues/1169